File: native-macos/CrossDashboard/UI/Screens/Memos/MemosViewModel.swift
import Foundation
import Observation
import CrossDashboardKit

@Observable
@MainActor
final class MemosViewModel {

    // ─── Dependencies ─────────────────────────────────────────────────────────

    private let container: AppContainer

    var memoRepo: MemoRepository { container.memoRepository }

    // ─── Filter state ─────────────────────────────────────────────────────────

    var stateFilter: MemoState? = .normal    // nil = all
    var selectedTags: Set<String> = []
    var searchText: String = ""

    // ─── Memo data ────────────────────────────────────────────────────────────

    var isLoading: Bool = false
    var error: String? = nil
    var snackbarMessage: String? = nil

    /// Comments keyed by memo name — loaded lazily.
    var comments: [String: [MemosMemo]] = [:]
    var commentLoading: Set<String> = []

    var filteredMemos: [MemosMemo] {
        var all = memoRepo.allMemos
        if let state = stateFilter {
            all = all.filter { $0.state == state }
        }
        all = all.filter { memo in selectedTags.allSatisfy { memo.tags.contains($0) } }
        if !searchText.isEmpty {
            all = all.filter { $0.content.localizedCaseInsensitiveContains(searchText) || $0.snippet.localizedCaseInsensitiveContains(searchText) }
        }
        return all
    }

    var allTags: [String] {
        Array(Set(memoRepo.allMemos.flatMap { $0.tags })).sorted()
    }

    func clearFilters() {
        stateFilter = .normal
        selectedTags = []
    }

    var memosHost: String {
        container.keychain.get(CredentialKey.memosHost) ?? ""
    }

    var memosToken: String {
        container.keychain.get(CredentialKey.memosToken) ?? ""
    }

    var configuredRepos: [String] {
        guard let raw = container.keychain.get(CredentialKey.giteaRepos) else { return [] }
        // SettingsViewModel.saveGitea() encodes repos as a JSON array — decode that here.
        if let data = raw.data(using: .utf8),
           let arr = try? JSONDecoder().decode([String].self, from: data) {
            return arr.filter { !$0.isEmpty }
        }
        // Fallback: plain comma-separated (e.g. Android-side format or manual entry)
        return raw.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
    }

    /// Open issues cached locally for a given repo, used by CommentOnIssueSheet.
    func issues(for repo: String) -> [GiteaIssue] {
        container.issueRepository.allIssues.filter { $0.repository == repo && $0.state == "open" }
    }

    // ─── Init ─────────────────────────────────────────────────────────────────

    init(container: AppContainer = .shared) {
        self.container = container
    }

    // ─── Sync ─────────────────────────────────────────────────────────────────

    func sync() async {
        isLoading = true
        defer { isLoading = false }
        await memoRepo.syncMemos()
    }

    // ─── Create ───────────────────────────────────────────────────────────────

    func createMemo(content: String, visibility: MemoVisibility, attachments: [MemoPendingAttachment]) async {
        guard !content.isEmpty else { return }
        isLoading = true
        defer { isLoading = false }
        if await memoRepo.createMemo(content: content, visibility: visibility, attachments: attachments) != nil {
            snackbarMessage = "Memo created"
        } else {
            error = "Failed to create memo"
        }
    }

    // ─── Delete / Archive ─────────────────────────────────────────────────────

    func deleteMemo(_ name: String, force: Bool = false) async {
        await memoRepo.deleteMemo(name, force: force)
        snackbarMessage = "Memo deleted"
    }

    func archiveMemo(_ name: String) async {
        await memoRepo.archiveMemo(name)
        snackbarMessage = "Memo archived"
    }

    func restoreMemo(_ name: String) async {
        await memoRepo.restoreMemo(name)
        snackbarMessage = "Memo restored"
    }

    // ─── Comments ─────────────────────────────────────────────────────────────

    func loadComments(for memoName: String) async {
        guard !commentLoading.contains(memoName) else { return }
        commentLoading.insert(memoName)
        defer { commentLoading.remove(memoName) }
        let result = await memoRepo.loadComments(memoName)
        comments[memoName] = result
    }

    func addComment(to memoName: String, content: String) async {
        guard !content.isEmpty else { return }
        _ = await memoRepo.createComment(parentId: memoName, content: content)
        await loadComments(for: memoName)
    }

    // ─── Share ────────────────────────────────────────────────────────────────

    func createShare(_ memoId: String) async -> String? {
        await memoRepo.createShare(memoId)
    }

    // ─── Action helpers ───────────────────────────────────────────────────────

    func extractTasks(from memo: MemosMemo) -> [ParsedTask] {
        return memo.content.components(separatedBy: "\n")
            .filter { $0.trimmingCharacters(in: .whitespaces).hasPrefix("- [ ]") }
            .compactMap { line -> ParsedTask? in
                let raw = line.trimmingCharacters(in: .whitespaces)
                    .replacingOccurrences(of: "- [ ]", with: "")
                    .trimmingCharacters(in: .whitespaces)
                guard !raw.isEmpty else { return nil }
                return TaskInputParser.parse(input: raw)
            }
    }

    func detectFirstDate(in memo: MemosMemo) -> Date? {
        let lower = memo.content.lowercased()
        if lower.contains("today")    { return Date() }
        if lower.contains("tomorrow") { return Calendar.current.date(byAdding: .day, value: 1, to: Date()) }
        // ISO date
        let regex = try? NSRegularExpression(pattern: #"\d{4}-\d{2}-\d{2}"#)
        if let match = regex?.firstMatch(in: memo.content, range: NSRange(memo.content.startIndex..., in: memo.content)),
           let range = Range(match.range, in: memo.content) {
            let dateStr = String(memo.content[range])
            let fmt = DateFormatter(); fmt.dateFormat = "yyyy-MM-dd"
            return fmt.date(from: dateStr)
        }
        return nil
    }

    func firstURL(in memo: MemosMemo) -> URL? {
        let detector = try? NSDataDetector(types: NSTextCheckingResult.CheckingType.link.rawValue)
        let range = NSRange(memo.content.startIndex..., in: memo.content)
        return detector?.firstMatch(in: memo.content, range: range)
            .flatMap { Range($0.range, in: memo.content) }
            .flatMap { URL(string: String(memo.content[$0])) }
    }

    func memoURL(for memo: MemosMemo) -> URL? {
        guard !memosHost.isEmpty else { return nil }
        return URL(string: "\(memosHost)/\(memo.name)")
    }

    // ─── Task / Event / Issue helpers ─────────────────────────────────────────

    func createTaskFromParsed(_ parsed: ParsedTask, calendarHref: String) async {
        let task = CalDavTask(summary: parsed.summary, priority: parsed.priority, due: parsed.due, categories: parsed.categories)
        _ = try? await container.taskRepository.create(task, calendarHref: calendarHref)
    }

    func createEventFromMemo(summary: String, description: String, start: Date, end: Date, calendarHref: String) async {
        let event = CalendarEvent(uid: UUID().uuidString, summary: summary, start: start, end: end, description: description)
        _ = try? await container.eventRepository.create(event, calendarHref: calendarHref)
        snackbarMessage = "Event created"
    }

    func addCommentToIssue(repo: String, issueNumber: Int, body: String) async {
        _ = try? await container.issueRepository.addComment(repo: repo, number: issueNumber, body: body)
        snackbarMessage = "Comment added"
    }

    func clearSnackbar() { snackbarMessage = nil }
    func clearError() { error = nil }
}

File: native-android/app/src/main/kotlin/com/crossdashboard/app/ui/screen/memos/MemosViewModel.kt
package com.crossdashboard.app.ui.screen.memos

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.crossdashboard.app.data.parser.TaskInputParser
import com.crossdashboard.app.data.prefs.CredentialKey
import com.crossdashboard.app.data.prefs.SecureStore
import com.crossdashboard.app.data.repository.EventRepository
import com.crossdashboard.app.data.repository.IssueRepository
import com.crossdashboard.app.data.repository.MemoRepository
import com.crossdashboard.app.data.repository.PendingAttachment
import com.crossdashboard.app.data.repository.TaskRepository
import com.crossdashboard.app.domain.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import javax.inject.Inject

enum class MemoStateFilter { NORMAL, ARCHIVED, ALL }

data class MemosUiState(
    val memos: List<MemosMemo> = emptyList(),
    val stateFilter: MemoStateFilter = MemoStateFilter.NORMAL,
    val availableTags: List<String> = emptyList(),
    val selectedTags: Set<String> = emptySet(),
    val dateRangeStart: Instant? = null,
    val dateRangeEnd: Instant? = null,
    val isLoading: Boolean = false,
    val error: String? = null,
    val comments: Map<String, List<MemosMemo>> = emptyMap(),   // keyed by memo name
    val commentLoading: Set<String> = emptySet(),
    val snackbarMessage: String? = null,
    val configuredRepos: List<String> = emptyList(),
    val memosHost: String = "",
    val memosToken: String = "",
    /** Open issues from Room, grouped by "owner/repo", used by CommentOnIssueSheet. */
    val issuesByRepo: Map<String, List<GiteaIssue>> = emptyMap(),
)

@HiltViewModel
class MemosViewModel @Inject constructor(
    private val memoRepo: MemoRepository,
    private val taskRepo: TaskRepository,
    private val eventRepo: EventRepository,
    private val issueRepo: IssueRepository,
    private val secureStore: SecureStore,
) : ViewModel() {

    private val _state = MutableStateFlow(MemosUiState())
    val state: StateFlow<MemosUiState> = _state.asStateFlow()

    init {
        val host  = secureStore.get(CredentialKey.MEMOS_HOST)  ?: ""
        val token = secureStore.get(CredentialKey.MEMOS_TOKEN) ?: ""
        val repos = secureStore.get(CredentialKey.GITEA_REPOS)
            ?.split(",")?.map { it.trim() }?.filter { it.isNotBlank() }
            ?: emptyList()
        _state.update { it.copy(memosHost = host, memosToken = token, configuredRepos = repos) }

        viewModelScope.launch {
            memoRepo.allMemos.collect { all ->
                applyFilters(all)
            }
        }

        viewModelScope.launch {
            issueRepo.allIssues.collect { all ->
                _state.update { it.copy(
                    issuesByRepo = all.filter { issue -> issue.state == "open" }
                        .groupBy { issue -> issue.repository }
                ) }
            }
        }
    }

    private fun applyFilters(all: List<MemosMemo>) {
        val s = _state.value
        var filtered = when (s.stateFilter) {
            MemoStateFilter.NORMAL   -> all.filter { it.state == MemoState.NORMAL }
            MemoStateFilter.ARCHIVED -> all.filter { it.state == MemoState.ARCHIVED }
            MemoStateFilter.ALL      -> all
        }
        filtered = filtered.filter { memo -> s.selectedTags.all { it in memo.tags } }
        s.dateRangeStart?.let { start -> filtered = filtered.filter { it.displayTime >= start } }
        s.dateRangeEnd?.let { end -> filtered = filtered.filter { it.displayTime <= end } }
        _state.update {
            it.copy(
                memos = filtered,
                availableTags = all.flatMap { memo -> memo.tags }
                    .distinct().sortedBy { tag -> tag.lowercase() },
            )
        }
    }

    fun setStateFilter(filter: MemoStateFilter) {
        _state.update { it.copy(stateFilter = filter) }
        viewModelScope.launch {
            applyFilters(memoRepo.allMemos.first())
        }
    }

    fun setTagFilters(tags: Set<String>) {
        _state.update { it.copy(selectedTags = tags) }
        viewModelScope.launch {
            applyFilters(memoRepo.allMemos.first())
        }
    }

    fun clearFilters() {
        _state.update {
            it.copy(stateFilter = MemoStateFilter.NORMAL, selectedTags = emptySet())
        }
        viewModelScope.launch { applyFilters(memoRepo.allMemos.first()) }
    }

    fun setDateRange(start: Instant?, end: Instant?) {
        _state.update { it.copy(dateRangeStart = start, dateRangeEnd = end) }
        viewModelScope.launch {
            applyFilters(memoRepo.allMemos.first())
        }
    }

    fun sync() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, error = null) }
            runCatching { memoRepo.syncMemos() }
                .onFailure { e -> _state.update { it.copy(error = e.message) } }
            _state.update { it.copy(isLoading = false) }
        }
    }

    fun createMemo(
        content: String,
        visibility: MemoVisibility,
        attachments: List<PendingAttachment>,
    ) {
        if (content.isBlank()) return
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, error = null) }
            val created = runCatching {
                memoRepo.createMemo(content, visibility, attachments)
            }.getOrNull()
            if (created != null) {
                _state.update { it.copy(snackbarMessage = "Memo created") }
            } else {
                _state.update { it.copy(error = "Failed to create memo") }
            }
            _state.update { it.copy(isLoading = false) }
        }
    }

    fun deleteMemo(name: String, force: Boolean = false) {
        viewModelScope.launch {
            memoRepo.deleteMemo(name, force)
            _state.update { it.copy(snackbarMessage = "Memo deleted") }
        }
    }

    fun archiveMemo(name: String) {
        viewModelScope.launch {
            memoRepo.archiveMemo(name)
            _state.update { it.copy(snackbarMessage = "Memo archived") }
        }
    }

    fun restoreMemo(name: String) {
        viewModelScope.launch {
            memoRepo.restoreMemo(name)
            _state.update { it.copy(snackbarMessage = "Memo restored") }
        }
    }

    fun loadComments(memoName: String) {
        if (_state.value.commentLoading.contains(memoName)) return
        viewModelScope.launch {
            _state.update { it.copy(commentLoading = it.commentLoading + memoName) }
            val comments = runCatching { memoRepo.loadComments(memoName) }.getOrDefault(emptyList())
            _state.update {
                it.copy(
                    comments = it.comments + (memoName to comments),
                    commentLoading = it.commentLoading - memoName,
                )
            }
        }
    }

    fun createComment(parentName: String, content: String) {
        if (content.isBlank()) return
        viewModelScope.launch {
            runCatching { memoRepo.createComment(parentName, content) }
            loadComments(parentName)
        }
    }

    fun createShare(memoId: String, onResult: (String?) -> Unit) {
        viewModelScope.launch {
            val url = memoRepo.createShare(memoId)
            onResult(url)
        }
    }

    // ─── Action helpers ───────────────────────────────────────────────────────

    /** Extract markdown task items from memo content. */
    fun extractTasks(memo: MemosMemo): List<ParsedTask> {
        return memo.content
            .lines()
            .filter { it.trimStart().startsWith("- [ ]") }
            .mapNotNull { line ->
                val raw = line.trimStart().removePrefix("- [ ]").trim()
                if (raw.isNotBlank()) TaskInputParser.parse(raw) else null
            }
    }

    /** Detect dates/keywords in memo content for event creation seeding. */
    fun detectFirstDate(memo: MemosMemo): Instant? {
        val keywords = mapOf(
            "today" to 0L, "tomorrow" to 1L,
            "monday" to null, "tuesday" to null, "wednesday" to null,
            "thursday" to null, "friday" to null, "saturday" to null, "sunday" to null,
        )
        val lower = memo.content.lowercase()
        for ((kw, offset) in keywords) {
            if (lower.contains(kw)) {
                return if (offset != null) {
                    Instant.now().plusSeconds(offset * 86400)
                } else {
                    Instant.now()
                }
            }
        }
        // ISO date: 2025-12-25
        val isoRegex = Regex("""\d{4}-\d{2}-\d{2}""")
        isoRegex.find(memo.content)?.let {
            return runCatching {
                java.time.LocalDate.parse(it.value)
                    .atStartOfDay(ZoneId.systemDefault())
                    .toInstant()
            }.getOrNull()
        }
        return null
    }

    /** Return first URL found in memo content, or null. */
    fun firstUrl(memo: MemosMemo): String? {
        val urlRegex = Regex("""https?://\S+""")
        return urlRegex.find(memo.content)?.value
    }

    fun createTaskFromParsed(parsed: ParsedTask) {
        viewModelScope.launch {
            val calendarHref = secureStore.get(CredentialKey.CALDAV_DEFAULT_TASK_CALENDAR) ?: return@launch
            runCatching {
                taskRepo.create(
                    CalDavTask(
                        uid = java.util.UUID.randomUUID().toString(),
                        summary = parsed.summary,
                        priority = parsed.priority,
                        categories = parsed.categories,
                        due = parsed.due,
                    ),
                    calendarHref,
                )
            }
        }
    }

    fun createEventFromMemo(
        summary: String,
        description: String,
        start: Instant,
        end: Instant,
        calendarHref: String,
    ) {
        viewModelScope.launch {
            runCatching {
                eventRepo.create(
                    CalendarEvent(
                        uid = java.util.UUID.randomUUID().toString(),
                        summary = summary,
                        start = start,
                        end = end,
                        description = description,
                    ),
                    calendarHref,
                )
            }.onSuccess { _state.update { it.copy(snackbarMessage = "Event created") } }
                .onFailure { e -> _state.update { it.copy(error = e.message) } }
        }
    }

    fun addCommentToIssue(repo: String, issueNumber: Int, body: String) {
        viewModelScope.launch {
            runCatching { issueRepo.addComment(repo, issueNumber, body) }
                .onSuccess { _state.update { it.copy(snackbarMessage = "Comment added") } }
                .onFailure { e -> _state.update { it.copy(error = e.message) } }
        }
    }

    fun clearSnackbar() = _state.update { it.copy(snackbarMessage = null) }
    fun showSnackbar(msg: String) = _state.update { it.copy(snackbarMessage = msg) }
    fun clearError() = _state.update { it.copy(error = null) }
}
